rust/otap-dataflow/crates/telemetry/src/opentelemetry_client/meter_provider.rs

```diff
@@ -0,0 +1,300 @@
+        reader_config: &MetricsReaderConfig,
+        mut runtime: Option<tokio::runtime::Runtime>,
+    ) -> Result<(MeterProviderBuilder, Option<tokio::runtime::Runtime>), Error> {
+        match reader_config {
+            MetricsReaderConfig::Periodic(periodic_config) => {
+                let interval = &periodic_config.interval;
+                match &periodic_config.exporter {
+                    MetricsPeriodicExporterConfig::Console => {
+                        sdk_meter_builder =
+                            Self::configure_console_metric_exporter(sdk_meter_builder, interval)?;
+                    }
+                    MetricsPeriodicExporterConfig::Otlp(otlp_config) => {
+                        (sdk_meter_builder, runtime) = Self::configure_otlp_metric_exporter(
+                            sdk_meter_builder,
+                            otlp_config,
+                            interval,
+                            runtime,
+                        )?;
+                    }
+                }
+                Ok((sdk_meter_builder, runtime))
+            }
+            MetricsReaderConfig::Pull { .. } => Err(Error::ConfigurationError(
+                "Pull-based metric readers are not implemented yet".to_string(),
+            )),
+        }
+    }
+
+    fn configure_otlp_metric_exporter(
+        mut sdk_meter_builder: MeterProviderBuilder,
+        otlp_config: &OtlpExporterConfig,
+        interval: &std::time::Duration,
+        mut runtime: Option<tokio::runtime::Runtime>,
+    ) -> Result<(MeterProviderBuilder, Option<tokio::runtime::Runtime>), Error> {
+        let exporter;
+        match &otlp_config.protocol {
+            OtlpProtocol::Grpc => {
+                (exporter, runtime) = Self::configure_grpc_otlp_exporter(otlp_config, runtime)?
+            }
+            OtlpProtocol::HttpBinary => {
+                exporter = Self::configure_http_exporter(otlp_config, Protocol::HttpBinary)?
+            }
+            OtlpProtocol::HttpJson => {
+                exporter = Self::configure_http_exporter(otlp_config, Protocol::HttpJson)?
+            }
+        };
+        let reader = PeriodicReader::builder(exporter)
+            .with_interval(*interval)
+            .build();
+        sdk_meter_builder = sdk_meter_builder.with_reader(reader);
+
+        //sdk_meter_builder = sdk_meter_builder.with_periodic_exporter(exporter);
+        Ok((sdk_meter_builder, runtime))
+    }
+
+    fn configure_grpc_otlp_exporter(
+        otlp_config: &OtlpExporterConfig,
+        runtime: Option<tokio::runtime::Runtime>,
+    ) -> Result<
+        (
+            opentelemetry_otlp::MetricExporter,
+            Option<tokio::runtime::Runtime>,
+        ),
+        Error,
+    > {
+        // If there is a tokio runtime already, use it. Otherwise, create a new one.
+        let tokio_runtime = match runtime {
+            Some(rt) => rt,
+            None => tokio::runtime::Runtime::new()
```

Comment:
This part , I am not sure what is the right thing to do. If we just create a runtime like above, will it create new thread(s)? Or should we use the current-thread runtime? I am unsure how this would affect rest of df_engine.

Reply:
I also opened https://github.com/open-telemetry/otel-arrow/issues/1573 to discuss this more. Maybe there is no action required, but something to track.